src/drivers/actuators/vertiq_io/entry_wrapper.hpp

```diff
@@ -0,0 +1,131 @@
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+class AbstractEntryWrapper
+{
+public:
+	AbstractEntryWrapper() {}
+	virtual void ConfigureStruct(param_t parameter, ClientEntryAbstract *entry) = 0;
+	virtual void SetNeedsInit() = 0;
+	virtual ClientEntryAbstract *GetClientEntry() = 0;
+	virtual void SendGet(VertiqSerialInterface *ser) = 0;
+	virtual void Update(VertiqSerialInterface *serial_interface) = 0;
+};
+
+template <typename module_data_type, typename px4_data_type>
+class EntryWrapper : public AbstractEntryWrapper
+{
+public:
+
+	param_t _param;
+	ClientEntry<module_data_type> *_entry;
+	bool _needs_init;
+
+	px4_data_type _px4_value;
+
+	EntryWrapper() {}
+
+	void ConfigureStruct(param_t parameter, ClientEntryAbstract *entry)
+	{
+		_param = parameter;
+		_entry = (ClientEntry<module_data_type> *)entry;
+		_needs_init = true;
+	}
+
+	void SetNeedsInit()
+	{
+		_needs_init = true;
+	}
+
+	ClientEntryAbstract *GetClientEntry()
+	{
+		return _entry;
+	}
+
+	void SendGet(VertiqSerialInterface *ser)
+	{
+		_entry->get(*ser->get_iquart_interface());
+		ser->process_serial_tx();
+	}
+
+	void SendSetAndSave(VertiqSerialInterface *ser, module_data_type value)
+	{
+		_entry->set(*ser->get_iquart_interface(), value);
+		_entry->save(*ser->get_iquart_interface());
+		ser->process_serial_tx();
+	}
+
+	bool ValuesAreTheSame(int32_t val1, int32_t val2, int32_t tolerance)
+	{
+		return (val1 == val2);
+	}
+
+	bool ValuesAreTheSame(float val1, float val2, float tolerance)
+	{
+		float diff = val1 - val2;
+
+		if(diff < 0){
+			return -diff < tolerance;
+		}
+
+		return diff < tolerance;
+	}
+
+	void Update(VertiqSerialInterface *serial_interface)
+	{
+		//Make sure we have a way to store the data from PX4 and from the module
+		module_data_type module_value = 0;
+		px4_data_type px4_value = 0;
+
+		//If there's new data in the entry
+		if (_entry->IsFresh()) {
+			//Grab the data
+			module_value = _entry->get_reply();
+
+			//If we need to set the PX4 data to the module's value, otherwise set the module value
+			if (_needs_init) {
+				px4_value = (px4_data_type)module_value;
+				param_set(_param, &px4_value);
+				_needs_init = false;
+
+			} else {
+				//Grab the PX4 value
```

Comment:
When do we actually get into this else logic? The section where "_needs_init" is true I get, that would trigger on a refresh. What's the scenario where are we using this portion, where PX4 has the latest value and needs to send it to the module? Is this how we update the modules when we change a parameter on the PX4 side?

Reply:
The else logic gets triggered when you set a module value through px4. Generally, that means changing a value through qgroundcontrol, but you could do it through mavlink as well. There's only two things you can do between px4 and module params. You can either set PX4 to the module value (_needs_init) or you can set the module value to the PX4 value (the else).